drivers/tee/optee/optee.c

```diff
@@ -1126,23 +1127,94 @@ static int set_optee_method(const struct device *dev)
 	return 0;
 }
 
+static bool optee_check_uid(const struct device *dev)
+{
+	struct arm_smccc_res res;
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+
+	data->smc_call(OPTEE_SMC_CALLS_UID, 0, 0, 0, 0, 0, 0, 0, &res);
+
+	if (res.a0 == OPTEE_MSG_UID_0 && res.a1 == OPTEE_MSG_UID_1 &&
+	    res.a2 == OPTEE_MSG_UID_2 && res.a3 == OPTEE_MSG_UID_3) {
+		return true;
+	}
+
+	return false;
+}
+
+static void optee_get_revision(const struct device *dev)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	union {
+		struct arm_smccc_res smccc;
+		struct optee_smc_call_get_os_revision_result result;
+	} res = { 0 };
+
+	data->smc_call(OPTEE_SMC_CALL_GET_OS_REVISION, 0, 0, 0, 0, 0, 0, 0,
+		  &res.smccc);
+
+	if (res.result.build_id) {
+		LOG_INF("OPTEE revision %lu.%lu (%08lx)", res.result.major,
```

Comment:
It is unsafe to use union in this way. You are basically introducing unspecified behavior there. C11 standard says that you can read only those values from union which you previously wrote to.